### PR TITLE
Block until FunctionEnvironmentReload request completes

### PR DESCRIPTION
**Pull request by pragnagopa**
Race condition:
Sometimes WorkerProcessReadyEvent is getting fired immediately even before language worker specialization begins to wait. This results in a Timeout Exception as event handler for WorkerProcessReadyEvent never gets invoked.

### CODE REVIEW

**Comment by pragnagopa** [comment]
cc @maiqbal11 @anirudhgarg @Hazhzeng

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 1]
```diff
@@ -94,13 +94,9 @@ public async Task SpecializeAsync()
...
                 _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
-                IObservable<WorkerProcessReadyEvent> processReadyEvents = _eventManager.OfType<WorkerProcessReadyEvent>()
-                .Where(msg => string.Equals(msg.Language, _workerRuntime, StringComparison.OrdinalIgnoreCase))
-                .Timeout(workerInitTimeout);
-                languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
-
-                // Wait for response from language worker process
-                await processReadyEvents.FirstAsync();
+                TaskCompletionSource<bool> reloadTask = new TaskCompletionSource<bool>();
+                languageWorkerChannel.SendFunctionEnvironmentReloadRequest(reloadTask);
```
Comment (fabiocav) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Why is this being passed from the consumer? any reason why this token/task isn't fully managed by the channel instead?

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: SendFunctionEnvironmentReloadRequest is called during specialization. Caller of the method needs to wait until FunctionEnvironmentReloadResponse is received from the language worker. Let me know if there is a better way to achieve this.

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 2]
```diff
@@ -206,10 +207,14 @@ public void SendFunctionLoadRequests()
             }
         }
 
-        public void SendFunctionEnvironmentReloadRequest()
+        public void SendFunctionEnvironmentReloadRequest(TaskCompletionSource<bool> reloadTask)
         {
-            _eventSubscriptions.Add(_inboundWorkerEvents.Where(msg => msg.MessageType == MsgType.FunctionEnvironmentReloadResponse)
-      .Subscribe((msg) => PublishWorkerProcessReadyEvent(msg.Message.FunctionEnvironmentReloadResponse)));
+            _workerChannelLogger.LogDebug("Sending FunctionEnvironmentReloadRequest");
```
Comment (fabiocav) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: This design assumes a single call and doesn't enforce that. The completion should be scoped to this request and managed by this type.

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 1 above)
Comment (fabiocav) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: This shouldn't be manged by the consumer. Also left a comment in the method implementation as the new approach is a bit fragile and opens up the opportunity for other potential issues. This should be managed by the worker channel

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs (hunk 2 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: thanks. Makes sense. Pushed an update

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 3]
```diff
@@ -223,6 +228,7 @@ public void SendFunctionEnvironmentReloadRequest()
             {
                 FunctionEnvironmentReloadRequest = request
             });
+            return _reloadTask.Task;
```
Comment (fabiocav) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: This approach is cleaner/better than what we had before.

With this design, if this is called more than once (I realize that is unexpected), we would be in a bad state, with a potentially difficult issue to troubleshoot/diagnose. If we don't expect that to be the case, we should likely throw from here if a reload is called again

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: updated and added test! Thanks!

Comment (fabiocav) [reply src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: Could we improve the logic to perform the validation when this method is called, as opposed to when the response comes back? That would ensure we're protecting the workers if this condition ever happens and would also avoid unnecessary work associated with the message handling.

**Comment by fabiocav** [comment]
Merged with c2825fc569a997cafd59bcf546b29ca69e9931a2

**Pull request closed without merging**